src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -0,0 +1,664 @@
+#include "mpidu_shm.h"
+#include "mpidu_shm_seg.h"
+
+#include <stdlib.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#include <errno.h>
+
+#if defined (HAVE_SYSV_SHARED_MEM)
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#endif
+
+#include <stddef.h>
+
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_SHM_RANDOM_ADDR_RETRY
+      category    : MEMORY
+      type        : int
+      default     : 100
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_GROUP
+      description : >-
+        The default number of retries for generating a random address. A retrying
+        involves only local operations.
+
+    - name        : MPIR_CVAR_SHM_SYMHEAP_RETRY
+      category    : MEMORY
+      type        : int
+      default     : 100
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_GROUP
+      description : >-
+        The default number of retries for allocating a symmetric heap in shared
+        memory. A retrying involves collective communication over the group in
+        the shared memory.
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+enum {
+    MPIDU_SYMSHM_SUCCESS,
+    MPIDU_SYMSHM_MAP_FAIL,      /* mapping failure when specified with a fixed start addr */
+    MPIDU_SYMSHM_OTHER_FAIL     /* other failure reported by MPL shm */
+};
+
+/* Linked list internally used to keep track
+ * of allocate shared memory segments */
+typedef struct seg_list {
+    void *key;
+    MPIDU_shm_seg_t *shm_seg;
+    struct seg_list *next;
+} seg_list_t;
+
+static seg_list_t *seg_list_head = NULL;
+static seg_list_t *seg_list_tail = NULL;
+
+size_t MPIDU_shm_get_mapsize(size_t size, size_t * psz)
+{
+    size_t page_sz, mapsize;
+
+    if (*psz == 0)
+        page_sz = (size_t) sysconf(_SC_PAGESIZE);
+    else
+        page_sz = *psz;
+
+    mapsize = (size + (page_sz - 1)) & (~(page_sz - 1));
+    *psz = page_sz;
+
+    return mapsize;
+}
+
+#ifdef USE_SYM_HEAP
+static int check_maprange_ok(void *start, size_t size)
+{
+    int rc = 0;
+    int ret = 0;
+    size_t page_sz = 0;
+    size_t mapsize = MPIDU_shm_get_mapsize(size, &page_sz);
+    size_t i, num_pages = mapsize / page_sz;
+    char *ptr = (char *) start;
+
+    for (i = 0; i < num_pages; i++) {
+        rc = msync(ptr, page_sz, 0);
+
+        if (rc == -1) {
+            if (errno != ENOMEM)
+                goto fn_fail;
+            ptr += page_sz;
+        } else
+            goto fn_exit;
+    }
+
+  fn_fail:
+    ret = 1;
+  fn_exit:
+    return ret;
+}
+
+static void *generate_random_addr(size_t size)
+{
+    /* starting position for pointer to map
+     * This is not generic, probably only works properly on Linux
+     * but it's not fatal since we bail after a fixed number of iterations
+     */
+#define MPIDU_MAP_POINTER ((random_unsigned&((0x00006FFFFFFFFFFF&(~(page_sz-1)))|0x0000600000000000)))
+    uintptr_t map_pointer;
```

Comment:
`MPIDU_MAP_POINTER` is only internally used in the c file. Can we get rid of `MPIDU_`?